src/math.cpp

```diff
@@ -355,83 +527,144 @@ void randomseed(const CallContext& ctx) {
-                    std::regex_match(s.value, pattern_hex) ||
-                    std::regex_match(s.value, pattern_exp)) {
-                    return Value(std::stoi(s.value, nullptr, 0));
-                } else {
-                    return Nil();
-                }
-            },
-            [](auto /*unused*/) -> Value { return Nil(); }},
-        x.raw());
+               overloaded{
+                   [](Number n) -> Value {
+                       if (std::modf(n.value, &n.value) == 0) {
+                           return Value(n.value);
+                       } else {
+                           return Nil();
+                       }
+                   },
+                   [](const String& s) -> Value {
+                       std::regex pattern_number(R"(\s*-?\d+)");
+                       std::regex pattern_hex(R"(\s*-?0[xX][\dA-Fa-f]+)");
+                       std::regex pattern_exp(R"(\s*-?\d+\.?\d*[eE]\d+)");
+
+                       if (std::regex_match(s.value, pattern_number) ||
+                           std::regex_match(s.value, pattern_hex) ||
+                           std::regex_match(s.value, pattern_exp)) {
+                           return Value(std::stoi(s.value, nullptr, 0));
+                       } else {
+                           return Nil();
+                       }
+                   },
+                   [](auto /*unused*/) -> Value { return Nil(); }},
+               x.raw())
+        .with_origin(origin);
 }
 
 auto type(const CallContext& ctx) -> Value {
+    auto origin = Origin(UnaryOrigin{
+        .val = make_owning<Value>(ctx.arguments().get(0)),
+        .location = ctx.call_location(),
+        .reverse = [](const Value& /*unused*/,
+                      const Value& /*unused*/) -> std::optional<SourceChangeTree> {
+            return std::nullopt; // only with the information of the type it is not reversable
+        }});
+
     auto x = ctx.arguments().get(0);
 
     if (x.is_number()) {
         Number num = std::get<Number>(x);
         double fraction = std::modf(num.value, &num.value);
 
-        return Value(fraction == 0.0 ? "integer" : "float");
+        return Value(fraction == 0.0 ? "integer" : "float").with_origin(origin);
     } else {
-        return Nil();
+        return Value(Nil()).with_origin(origin);
     }
 }
 
 auto ult(const CallContext& ctx) -> Value {
+    auto origin = Origin(BinaryOrigin{
+        .lhs = make_owning<Value>(ctx.arguments().get(0)),
+        .rhs = make_owning<Value>(ctx.arguments().get(1)),
+        .location = ctx.call_location(),
+        .reverse = [](const Value& /*unused*/, const Value& /*unused*/,
+                      const Value& /*unused*/) -> std::optional<SourceChangeTree> {
+            return std::nullopt;
+        }}); // Not reverseable because only an bool is avaible and with only the bool, its not
+             // possible to gain correctly the numbers back.
```

Comment:
```suggestion
        }}); // Not reverseable because only an bool is available and with only the bool, its not
             // possible to correctly get the numbers back.
```